**Universidad Internacional de La Rioja (UNIR) - Máster Universitario en Inteligencia Artificial - Procesamiento del Lenguaje Natural** 

***
Datos del alumno (Nombre y Apellidos): Pablo Alberto Duque Marín    

Fecha: 14 de abril del 2026
***

<span style="font-size: 20pt; font-weight: bold; color: #0098cd;">Trabajo: Caracterización de textos</span>

**Objetivos** 

Con esta actividad se tratará de que el alumno se familiarice con el manejo de la librería spacy, así como con los conceptos básicos de manejo de las técnicas NER

**Descripción**

En esta actividad debes procesar de forma automática un texto en lenguaje natural para detectar características básicas en el mismo, y para identificar y etiquetar las ocurrencias de conceptos como localización, moneda, empresas, etc.

En la primera parte del ejercicio se proporciona un código fuente a través del cual se lee un archivo de texto y se realiza un preprocesado del mismo. En esta parte el alumno tan sólo debe ejecutar y entender el código proporcionado.

En la segunda parte del ejercicio se plantean una serie de preguntas que deben ser respondidas por el alumno. Cada pregunta deberá responderse con un fragmento de código fuente que esté acompañado de la explicación correspondiente. Para elaborar el código solicitado, el alumno deberá visitar la documentación de la librería spacy, cuyos enlaces se proporcionarán donde corresponda.

# Parte 1: carga y preprocesamiento del texto a analizar

Observa las diferentes librerías que se están importando.

In [1]:
import pathlib
import spacy
import pandas as pd
from spacy import displacy
import csv
import es_core_news_md
import numpy as np

El siguiente código simplemente carga y preprocesa el texto. Para ello, lo primero que hace es cargar un modelo de lenguaje previamente entrenado. En este caso, se utiliza <i>es_core_news_md</i>: 

https://spacy.io/models/es#es_core_news_md


In [2]:
nlp = es_core_news_md.load()

El objeto <i>nlp</i> permite utilizar el modelo de lenguaje cargado, de forma que se puede procesar un texto y obtenerlo en su versión preprocesada. Así, nos permite realizar las diferentes tareas. En este caso, vamos a utilizar el pipeline para hacer un preprocesamiento básico, que consiste en tokenizar el texto.

In [3]:
filename = "./02Dataset_anonimizado.csv"
lines_number = 10000
data = pd.read_csv(filename, nrows=lines_number, sep=";", encoding="latin-1")

In [4]:
data.head()

,MEDIO,SOPORTE,URL,TIPO DE MENSAJE,CONTENIDO A ANALIZAR,INTENSIDAD,TIPO DE ODIO,TONO HUMORISTICO,MODIFICADOR,Unnamed: 9,Unnamed: 10,Unnamed: 11,Unnamed: 12,Unnamed: 13,Unnamed: 14,Unnamed: 15
0,EL PAÍS,WEB,URL_a4d7efc0,COMENTARIO,el barí§a nunca acaeza ante un segundo b ni an...,3.0,Otros,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,EL PAÍS,WEB,URL_a4d7efc0,COMENTARIO,el real madrid ha puesto punto y final a su an...,0.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,EL PAÍS,WEB,URL_54312d9e,COMENTARIO,cristina cifuentes podrí­a haber sido la presi...,3.0,Ideológico,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,EL PAÍS,WEB,URL_54312d9e,COMENTARIO,habrí­a que reabrir el caso. el supremo se ded...,3.0,Ideológico,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,EL PAÍS,WEB,URL_54312d9e,COMENTARIO,me parece un poco exagerado pedir más de tres ...,3.0,Ideológico,Si,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


El código anterior carga el archivo CSV (opcionalmente con un límite de líneas a leer) y genera la variable <i>data</i>, que contiene un Dataframe (https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.html) con los datos leídos del CSV.

Te vendrá bien conocer la siguiente documentación:
<ul>
    <li>https://spacy.io/api/doc</li>
    <li>https://spacy.io/api/token</li>
    <li>https://spacy.io/api/morphology#morphanalysis</li>
</ul>

### Playground

Utiliza este espacio para hacer pruebas y ensayos con las variables generadas con el código previo. A modo de ejemplo, se ofrece código que realiza las siguientes tareas: 


- leer un número dado de líneas del Dataframe y generar dos listas con los valores (se pueden leer directamente del DataFrame, se muestra el ejemplo como una opción más)
- procesar el texto de cada comentario


Para procesarlo, hay utilizar el objeto <i>nlp</i> y así obtener objetos de la clase <i>Doc</i> (https://spacy.io/api/doc)

Visita la documentación de dicha clase y experimenta probando las diferentes funciones y atributos 

In [5]:
# Puedes insertar aquí código de pruebas para experimentar con las diferentes funciones y atributos de 'doc'.
print(data["CONTENIDO A ANALIZAR"][1])
print(data["INTENSIDAD"][1])
doc = []
value = []

#con el bucle, generamos sendas listas con los comentarios ya parseados y con el valor de intensidad
for i in range(0, lines_number):
    
    #en un primer paso se parsea el comentario. En el segundo paso se añade el objeto a la lista
    tmp_doc = nlp(data["CONTENIDO A ANALIZAR"][i])
    doc.append(tmp_doc)
    
    #en un primer paso extrae el valor. En el segundo paso se añade el valor a la lista
    tmp_value = data["INTENSIDAD"][i]
    value.append(tmp_value)


#ejemplo de cómo recorrer un comentario palabra por palabra    
for token in doc[1]:
    print(token)

el real madrid ha puesto punto y final a su andadura en la copa del rey en el primer escalón. los de zidane han caí­do ante el alcoyano, de segunda b, a pesar de empezar ganando y jugar con un hombre menos en la prórroga. el técnico francés dispuso un equipo plagado de los menos habituales, con vinicius y mariano en ataque. ninguno de los dos logró crear ocasiones. fue militao el que marcó el gol del madrid, justo antes del descanso. en la segunda parte intentaron cerrar el partido, pero sin el colmillo suficiente y el modesto alcoyano aprovechó un córner para empatar el partido a cinco minutos para el final. el empate sentó como un jarro de agua frí­a a los blancos, que lo intentaron en el tiempo extra a falta de cinco minutos, el casanova consiguió el gol más importante de su vida, que vale la clasificación para octavos de la copa. el madrid de zidane queda apeado del torneo una vez más, por lo que el francés se quedará sin pelear por el único tí­tulo que no ha conseguido nunca. así­

In [6]:
texto_prueba = nlp("Los gatos de María corren por el parque de Bogotá")
for token in texto_prueba:
    print(f"{token.text:15} | POS: {token.pos_:8} | Lemma: {token.lemma_:15} | Stop: {token.is_stop} | Punct: {token.is_punct}")

Los             | POS: DET      | Lemma: el              | Stop: True | Punct: False
gatos           | POS: NOUN     | Lemma: gato            | Stop: False | Punct: False
de              | POS: ADP      | Lemma: de              | Stop: True | Punct: False
María           | POS: PROPN    | Lemma: María           | Stop: False | Punct: False
corren          | POS: VERB     | Lemma: correr          | Stop: False | Punct: False
por             | POS: ADP      | Lemma: por             | Stop: True | Punct: False
el              | POS: DET      | Lemma: el              | Stop: True | Punct: False
parque          | POS: NOUN     | Lemma: parque          | Stop: False | Punct: False
de              | POS: ADP      | Lemma: de              | Stop: True | Punct: False
Bogotá          | POS: PROPN    | Lemma: Bogotá          | Stop: False | Punct: False


In [7]:
for ent in texto_prueba.ents:
    print(f"{ent.text:20} | Label: {ent.label_:10} | Explicación: {spacy.explain(ent.label_)}")

María                | Label: PER        | Explicación: Named person or family.
parque de Bogotá     | Label: LOC        | Explicación: Non-GPE locations, mountain ranges, bodies of water


<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Pregunta 1.</span>
<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">¿Cuántos registros contiene el corpus?</span>

In [8]:
# Incluye aquí el código generado para poder responder a tu pregunta
print(f"La cantidad de registros cargados de nuestro corpus es de {len(data)} registros.")     
for columna, nulos in data.isnull().sum().items():
    print(f"La cantidad de valores nulos en la columna '{columna}' es de {nulos} registros.")

La cantidad de registros cargados de nuestro corpus es de 10000 registros.
La cantidad de valores nulos en la columna 'MEDIO' es de 0 registros.
La cantidad de valores nulos en la columna 'SOPORTE' es de 0 registros.
La cantidad de valores nulos en la columna 'URL' es de 0 registros.
La cantidad de valores nulos en la columna 'TIPO DE MENSAJE' es de 0 registros.
La cantidad de valores nulos en la columna 'CONTENIDO A ANALIZAR' es de 0 registros.
La cantidad de valores nulos en la columna 'INTENSIDAD' es de 0 registros.
La cantidad de valores nulos en la columna 'TIPO DE ODIO' es de 9252 registros.
La cantidad de valores nulos en la columna 'TONO HUMORISTICO' es de 9977 registros.
La cantidad de valores nulos en la columna 'MODIFICADOR' es de 9964 registros.
La cantidad de valores nulos en la columna 'Unnamed: 9' es de 10000 registros.
La cantidad de valores nulos en la columna 'Unnamed: 10' es de 10000 registros.
La cantidad de valores nulos en la columna 'Unnamed: 11' es de 10000 regi

In [9]:
filas_vacias = data[data.isnull().all(axis=1)]

print(f"La cantidad de registros completamente nulos es de {len(filas_vacias)} registros.")

La cantidad de registros completamente nulos es de 0 registros.


<b>Incluye aquí, debajo de la línea, la explicación de tu respuesta</b>
<hr> Se verificó la cantidad de registros cargados en el dataframe usando la funcion len, para ver el tamaño del dataframe ya que la función devuelve la cantidad de filas del dataframe, tambien se hubiera logrado con el atributo .shape[0] (para el caso específico data.shape[0]), se constató que la cantidad cargada fue de 10000 registros, esto coincide con valor definido en la variable que se uso como limite para la carga lines_number.
 Adicionalmente se verifico la cantidad de nulos por columna usando un bucle for y contando por columna con un data.isnull().sum(), también la existencia de filas completamente vacías filtrando con data.isnull().all(axis=1). 
 

<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Pregunta 2.</span>
<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">¿Cuántas palabras totales hay en los comentarios del corpus?</span>

In [10]:
# Incluye aquí el código generado para poder responder a tu pregunta
total_palabras = sum(1 for comentario in doc for token in comentario if token.is_alpha)
print(f"La cantidad total de palabras en los comentarios del corpus es de {total_palabras} palabras.")

La cantidad total de palabras en los comentarios del corpus es de 201427 palabras.


<b>Incluye aquí, debajo de la línea, la explicación de tu respuesta</b>
<hr>Para sacar el total de palabras de los comentarios, lo que se hizo fue recorrer uno por uno los objetos Doc que ya teníamos procesados en la lista doc usando spaCy, y ahí se filtraron los tokens para contar solo los que dieran verdadero en la condición token.is_alpha. Se eligió este atributo porque asegura que el token sea puramente letras, eso excluye los puntos, comas, numeros.
Para que el proceso no fuera pesado y no consumiera tanta memoria RAM, se usó una expresión generadora con la función sum(). Al final dio un total de 201427 palabras en los 10,000 comentarios cargados. Hay que decir que este método es bastante estricto, porque si alguien escribió algo como "100pre", el filtro lo deja afuera por tener números, pero para el ejercicio que es caracterizar el corpus, sirve para tener un conteo limpio de las palabras reales.
 

<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Pregunta 3.</span>
<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">¿Cuál el número promedio de palabras en cada comentario?</span>

In [11]:
# Incluye aquí el código generado para poder responder a tu pregunta
counts = [sum(1 for token in comentario if token.is_alpha) for comentario in doc]

promedio_general = sum(counts) / len(counts)


print(f"El número promedio de palabras por comentario es de {promedio_general:.2f} palabras.")
print(f"El número máximo de palabras en un comentario es de {max(counts)} palabras.")
print(f"El número mínimo de palabras en un comentario es de {min(counts)} palabras.")   
print(f"La mediana del número de palabras por comentario es de {np.median(counts)} palabras.")
print (f"la desviación estándar del número de palabras por comentario es de {np.std(counts):.2f} palabras.")

El número promedio de palabras por comentario es de 20.14 palabras.
El número máximo de palabras en un comentario es de 2421 palabras.
El número mínimo de palabras en un comentario es de 0 palabras.
La mediana del número de palabras por comentario es de 12.0 palabras.
la desviación estándar del número de palabras por comentario es de 63.58 palabras.


<b>Incluye aquí, debajo de la línea, la explicación de tu respuesta</b>
<hr>Para calcular el promedio de palabras por comentario, en lugar de obtener únicamente un valor general, se optó por construir una lista llamada counts que almacena el conteo individual de palabras de cada uno de los 10,000 comentarios. Se aplicó nuevamente el filtro token.is_alpha para mantener coherencia con el criterio de conteo definido en la pregunta anterior. Posteriormente, se calculó el promedio dividiendo la suma total de counts entre el número de registros, obteniendo un resultado de 20.14 palabras por comentario. Adicionalmente se calcularon otras medidas descriptivas con el fin de tener una visión mas completa de la distribución del corpues, ya que el promedio es muy suceptible a valores atípicos, el valor maximo registrado de palabras en un comentario fue de 2421 palabras, y el minimo fue de 0 (el cual anteriormente se habia corroborado que no tenia registros nulos), esto sugiere la posibilidad de tener registros que no contienen palabras alfabeticas, que pueden estar compuestos por emojis o caracteres alfanumericos. La mediana obtenida fue de 12, que es significativamente inferior  a la media, podemos ver que la distribución de los datos esta sesgada a la derecha por comentarios muy largos. Con esto nos podemos hacer una idea que el comentario tipico tiene al rededor de 12 palabras y que el valor del promedio se ve afectado por comentarios inusualmente largos. La sd del numero de palabras por comentario fue de 63.58, que revela la alta dispersión que tienen los datos, aproximadamente 3 veces mas que la media, con esto se puede observar la gran variabilidad que tiene el corpus en la extensión de sus comentarios.

 

<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Pregunta 4.</span>
<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Considerando dos grupos de comentarios (odio y no odio) ¿Cuál el número promedio de palabras en los comentarios de cada grupo?</span>

In [12]:
print(value)

[np.float64(3.0), np.float64(0.0), np.float64(3.0), np.float64(3.0), np.float64(3.0), np.float64(3.0), np.float64(4.0), np.float64(3.0), np.float64(4.0), np.float64(3.0), np.float64(3.0), np.float64(4.0), np.float64(0.0), np.float64(4.0), np.float64(4.0), np.float64(3.0), np.float64(0.0), np.float64(3.0), np.float64(0.0), np.float64(0.0), np.float64(3.0), np.float64(3.0), np.float64(4.0), np.float64(4.0), np.float64(4.0), np.float64(3.0), np.float64(4.0), np.float64(3.0), np.float64(3.0), np.float64(4.0), np.float64(3.0), np.float64(3.0), np.float64(3.0), np.float64(0.0), np.float64(3.0), np.float64(3.0), np.float64(4.0), np.float64(4.0), np.float64(3.0), np.float64(3.0), np.float64(3.0), np.float64(4.0), np.float64(0.0), np.float64(3.0), np.float64(4.0), np.float64(3.0), np.float64(4.0), np.float64(4.0), np.float64(4.0), np.float64(4.0), np.float64(4.0), np.float64(4.0), np.float64(4.0), np.float64(3.0), np.float64(3.0), np.float64(3.0), np.float64(3.0), np.float64(4.0), np.float64(4.

In [13]:
# Incluye aquí el código generado para poder responder a tu pregunta

palabras_odio = [c for c, v in zip(counts, value) if v > 0]
palabras_no_odio = [c for c, v in zip(counts, value) if v == 0]

# Calculamos los promedios directos (ya que estamos 100% seguros de que hay datos)
promedio_odio = sum(palabras_odio) / len(palabras_odio)
promedio_no_odio = sum(palabras_no_odio) / len(palabras_no_odio)

print(f"Promedio palabras en comentarios de odio: {promedio_odio}")
print(f"la mediana del número de palabras por comentario de odio es de {np.median(palabras_odio)} palabras.")
print(f"la desviación estándar del número de palabras por comentario de odio es de {np.std(palabras_odio):.2f} palabras.")
print(f"Promedio palabras en comentarios sin odio: {promedio_no_odio}")
print(f"la mediana del número de palabras por comentario sin odio es de {np.median(palabras_no_odio)} palabras.")
print(f"la desviación estándar del número de palabras por comentario sin odio es de {np.std(palabras_no_odio):.2f} palabras.")  

Promedio palabras en comentarios de odio: 21.204545454545453
la mediana del número de palabras por comentario de odio es de 11.5 palabras.
la desviación estándar del número de palabras por comentario de odio es de 73.14 palabras.
Promedio palabras en comentarios sin odio: 20.056852572416776
la mediana del número de palabras por comentario sin odio es de 12.0 palabras.
la desviación estándar del número de palabras por comentario sin odio es de 62.74 palabras.


<b>Incluye aquí, debajo de la línea, la explicación de tu respuesta</b>
<hr>Para calcular el promedio de palabras separado por grupo, se utilizó la función zip() para recorrer simultáneamente las listas counts (que almacena el conteo de palabras por comentario) y value (que contiene la intensidad de odio). Se aplicó el criterio del enunciado: si INTENSIDAD == 0, el comentario se clasifica como "no odio"; si INTENSIDAD > 0, se clasifica como "odio". Mediante list comprehensions se generaron dos listas separadas: palabras_odio y palabras_no_odio, cada una conteniendo los conteos de palabras correspondientes a su grupo.
Los resultados obtenidos fueron un promedio de 21.20 palabras para los comentarios con odio y de 20.06 palabras para los comentarios sin odio. La diferencia entre ambos grupos es de apenas 1.14 palabras. Al complementar con otras medidas descriptivas, se observa que las medianas son prácticamente iguales (11.5 para odio y 12.0 para no odio), lo que confirma que el usuario típico escribe una cantidad similar de palabras independientemente de si el contenido es de odio o no.
Sin embargo, la desviación estándar sí presenta una diferencia notable: 73.14 para odio frente a 62.74 para no odio. Esto indica que los comentarios de odio tienen mayor variabilidad en su extensión, es decir, hay mayor presencia de comentarios extremadamente cortos o extremadamente largos en comparación con el grupo sin odio. A pesar de esta diferencia en dispersión, la similitud en las medidas de tendencia central sugiere que la cantidad de palabras por sí sola probablemente no sería una característica con alto poder discriminatorio para clasificar mensajes de odio frente a mensajes neutros.

<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Pregunta 5.</span>
<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Considerando dos grupos de comentarios (odio y no odio) ¿Cuál es el número promedio de oraciones en los comentarios de cada grupo?</span>

In [14]:
counts_sents = [sum(1 for oracion in comentario.sents) for comentario in doc]

sents_odio = [c for c, v in zip(counts_sents, value) if v > 0]
sents_no_odio = [c for c, v in zip(counts_sents, value) if v == 0]

promedio_sents_odio = sum(sents_odio) / len(sents_odio)
promedio_sents_no_odio = sum(sents_no_odio) / len(sents_no_odio)


print(f"Promedio de oraciones en comentarios de odio: {promedio_sents_odio}")
print(f"la mediana del número de oraciones por comentario de odio es de {np.median(sents_odio)} oraciones.")
print(f"la desviación estándar del número de oraciones por comentario de odio es de {np.std(sents_odio):.2f} oraciones.")
print(f"Promedio de oraciones en comentarios sin odio: {promedio_sents_no_odio}")
print(f"la mediana del número de oraciones por comentario sin odio es de {np.median(sents_no_odio)} oraciones.")
print(f"la desviación estándar del número de oraciones por comentario sin odio es de {np.std(sents_no_odio):.2f} oraciones.")

Promedio de oraciones en comentarios de odio: 1.7700534759358288
la mediana del número de oraciones por comentario de odio es de 1.0 oraciones.
la desviación estándar del número de oraciones por comentario de odio es de 2.04 oraciones.
Promedio de oraciones en comentarios sin odio: 1.7796152183311715
la mediana del número de oraciones por comentario sin odio es de 1.0 oraciones.
la desviación estándar del número de oraciones por comentario sin odio es de 2.34 oraciones.


<b>Incluye aquí, debajo de la línea, la explicación de tu respuesta</b>
<hr>
Para calcular el promedio de oraciones por grupo, se construyó una lista counts_sents que almacena el conteo individual de oraciones de cada uno de los 10,000 comentarios. Para ello se utilizó el atributo .sents del objeto Doc, el cual es un generador que produce las oraciones que spaCy identificó durante el procesamiento del texto, basándose en los límites oracionales que el modelo aprendió a detectar. El conteo se realizó mediante sum(1 for oracion in comentario.sents) para cada comentario. Posteriormente, se aplicó la misma separación por grupo utilizada en la pregunta anterior, usando zip() y list comprehensions con el criterio INTENSIDAD == 0 para "no odio" e INTENSIDAD > 0 para "odio".
Los resultados muestran que los promedios de oraciones son prácticamente idénticos: 1.77 para odio y 1.78 para no odio, con una diferencia despreciable de 0.009 oraciones. Las medianas coinciden en 1.0 oración para ambos grupos, y las desviaciones estándar son muy similares (2.04 para odio y 2.34 para no odio). Esto indica que el usuario típico utiliza la misma cantidad de oraciones tanto en comentarios con odio como sin odio, lo cual sugiere que esta característica tampoco tendría un poder discriminatorio significativo para una tarea de clasificación.
 

<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Pregunta 6.</span>
<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Considerando dos grupos de comentarios (odio y no odio) ¿Cuál es el porcentaje de comentarios que contienen entidades NER en cada grupo?</span>

In [15]:
# Incluye aquí el código generado para poder responder a tu pregunta
tiene_entidades = [len(comentario.ents) > 0 for comentario in doc]

ner_odio = [tiene for tiene, v in zip(tiene_entidades, value) if v > 0]
ner_no_odio = [tiene for tiene, v in zip(tiene_entidades, value) if v == 0]


porcentaje_odio = (sum(ner_odio) / len(ner_odio)) * 100
porcentaje_no_odio = (sum(ner_no_odio) / len(ner_no_odio)) * 100

print(f"Porcentaje de comentarios con NER (Odio): {porcentaje_odio:.2f}%")
print(f"Porcentaje de comentarios con NER (No Odio): {porcentaje_no_odio:.2f}%")

Porcentaje de comentarios con NER (Odio): 37.57%
Porcentaje de comentarios con NER (No Odio): 33.91%


<b>Incluye aquí, debajo de la línea, la explicación de tu respuesta</b>
<hr>Para calcular el porcentaje de comentarios que contienen entidades nombradas (NER) en cada grupo, se generó primero una lista de valores booleanos indicando la presencia de al menos una entidad en cada uno de los 10,000 documentos mediante el atributo .ents del objeto Doc. En Python, los valores booleanos True y False se evalúan como 1 y 0 en operaciones aritméticas, por lo que al sumar la lista y dividir por su longitud se obtiene directamente la proporción de comentarios con entidades. Posteriormente, estos valores se segregaron por grupo utilizando la lista de intensidades y la función zip(), aplicando el mismo criterio de clasificación de las preguntas anteriores.Los resultados arrojan que el 37.57% de los comentarios de odio contienen entidades, frente a un 33.91% en el grupo de no odio. Si bien existe una diferencia de 3.66 puntos porcentuales a favor del grupo de odio, esta brecha sigue siendo estrecha. A diferencia de las características anteriores (cantidad de palabras y oraciones), aquí se observa una ligera tendencia: los comentarios con odio tienden a mencionar entidades nombradas (como nombres de personas, lugares u organizaciones) con una frecuencia levemente mayor, lo cual es coherente con la naturaleza del discurso de odio, que suele dirigirse hacia sujetos o grupos concretos. No obstante, la diferencia no parece lo suficientemente contundente como para separar ambos grupos con total claridad.
 

<span style="font-size: 16pt; font-weight: bold; color: #0098cd;">Plantea tus propias preguntas</span>

<span><b>Plantea al menos 4 características</b> del texto cuyo análisis permita una caracterización completa del texto. Puedes utilizar recomendaciones proporcionadas por la IA Generativa, si así lo deseas. Para cada una de las características planteadas, obtén valores separados para los grupos ODIO/NO-ODIO.</span>

<span>En la explicación aportada, deberás <b>explicar el significado de la característica planteada</b> así como la importancia de ésta en la caracterización del texto.</span>

<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Característica adicional 1.</span>


<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">[Característica adicional 1: Proporción de tokens en mayúsculas]</span>

In [16]:
# Incluye aquí el código generado para poder responder a tu pregunta
counts_upper = []
for comentario in doc:
    # Contamos cuántos tokens son palabras (alfabéticos)
    solo_palabras = [t for t in comentario if t.is_alpha]
    total_palabras = len(solo_palabras)
    
    if total_palabras > 0:
        # De esas palabras, ¿cuántas están en mayúsculas?
        mayus = sum(1 for t in solo_palabras if t.is_upper)
        counts_upper.append(mayus / total_palabras)
    else:
        counts_upper.append(0.0)
upper_odio = [c for c, v in zip(counts_upper, value) if v > 0]
upper_no_odio = [c for c, v in zip(counts_upper, value) if v == 0]


print(f"Odio - Promedio: {np.mean(upper_odio):.4f}, Mediana: {np.median(upper_odio)}, Desviación: {np.std(upper_odio):.4f}")
print(f"No Odio - Promedio: {np.mean(upper_no_odio):.4f}, Mediana: {np.median(upper_no_odio)}, Desviación: {np.std(upper_no_odio):.4f}")

Odio - Promedio: 0.0000, Mediana: 0.0, Desviación: 0.0000
No Odio - Promedio: 0.0001, Mediana: 0.0, Desviación: 0.0104


<b>Incluye aquí, debajo de la línea, la explicación de la característica propuesta y su motivación. Incluye también una explicación del código fuente aportado.</b>
<hr><b></b>
En esta etapa se analizó la hipótesis de que el uso de palabras completamente en mayúsculas podría ser un indicador de agresividad en los comentarios de odio. Para ello, se calculó la proporción de tokens que cumplen la condición token.is_upper sobre el total de palabras alfabéticas (token.is_alpha) de cada comentario, manejando los casos de comentarios sin palabras alfabéticas con un valor de 0.0 para evitar divisiones por cero.
Los resultados mostraron proporciones prácticamente nulas en ambos grupos: 0.0000 para odio y 0.0001 para no odio. Las medianas fueron 0.0 en ambos casos y las desviaciones estándar despreciables (0.0000 para odio y 0.0104 para no odio), lo que indica que el usuario típico no utiliza palabras completas en mayúsculas, independientemente de la intención del mensaje.

 

<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Característica adicional 2.</span>


<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Análisis de la proporción de signos de puntuación</span>

In [17]:
# Incluye aquí el código generado para poder responder a tu pregunta
counts_punct = [
    sum(1 for token in comentario if token.is_punct) / len(comentario) 
    if len(comentario) > 0 else 0.0 
    for comentario in doc
]

punct_odio = [p for p, v in zip(counts_punct, value) if v > 0]
punct_no_odio = [p for p, v in zip(counts_punct, value) if v == 0]

print(f"ODIO - Promedio: {np.mean(punct_odio):.4f}, Mediana: {np.median(punct_odio)}, SD: {np.std(punct_odio):.4f}")
print(f"NO ODIO - Promedio: {np.mean(punct_no_odio):.4f}, Mediana: {np.median(punct_no_odio)}, SD: {np.std(punct_no_odio):.4f}")

ODIO - Promedio: 0.1452, Mediana: 0.1259920634920635, SD: 0.1257
NO ODIO - Promedio: 0.1599, Mediana: 0.13333333333333333, SD: 0.1350


<b>Incluye aquí, debajo de la línea, la explicación de la característica propuesta y su motivación. Incluye también una explicación del código fuente aportado.</b>
<hr><b>Característica adicional 2: Análisis de la proporción de signos de puntuación
</b>En este análisis se midió la proporción de signos de puntuación (token.is_punct) sobre el total de tokens en cada comentario. La hipótesis planteada era que el uso excesivo de signos o su ausencia podría ayudar a discriminar el discurso de odio.
Los resultados muestran que los comentarios sin odio presentan una proporción de puntuación ligeramente superior (0.16) frente a los de odio (0.14). Las medianas también reflejan esta tendencia (0.13 vs 0.12), con desviaciones estándar muy similares (0.13 y 0.12 respectivamente).
Esta pequeña diferencia sugiere que el discurso de odio en este corpus no se caracteriza necesariamente por un uso exagerado de signos de puntuación, sino más bien por una estructura levemente más simple o menos puntuada que el discurso neutro. Al igual que en las métricas anteriores, la cercanía entre estos valores indica que la densidad de puntuación por sí sola no es un predictor fuerte para identificar el odio.

 

<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Característica adicional 3.</span>


<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Proporción de adjetivos </span>

In [18]:
# Incluye aquí el código generado para poder responder a tu pregunta
counts_adj = []
for comentario in doc:
    palabras_total = [t for t in comentario if t.is_alpha]
    n_palabras = len(palabras_total)
    
    if n_palabras > 0:
        n_adj = sum(1 for t in palabras_total if t.pos_ == "ADJ")
        counts_adj.append(n_adj / n_palabras)
    else:
        counts_adj.append(0.0)

# Separar por grupo
adj_odio = [a for a, v in zip(counts_adj, value) if v > 0]
adj_no_odio = [a for a, v in zip(counts_adj, value) if v == 0]

# Estadísticas para ambos grupos
print(f"ODIO - Promedio: {np.mean(adj_odio):.4f}, Mediana: {np.median(adj_odio)}, SD: {np.std(adj_odio):.4f}")
print(f"NO ODIO - Promedio: {np.mean(adj_no_odio):.4f}, Mediana: {np.median(adj_no_odio)}, SD: {np.std(adj_no_odio):.4f}")



ODIO - Promedio: 0.0959, Mediana: 0.058823529411764705, SD: 0.1434
NO ODIO - Promedio: 0.0805, Mediana: 0.045454545454545456, SD: 0.1327


<b>Incluye aquí, debajo de la línea, la explicación de la característica propuesta y su motivación. Incluye también una explicación del código fuente aportado.</b>
<hr><b>Característica adicional 3: Proporción de adjetivos 
</b>
</b>En este apartado se evaluó si el discurso de odio tiende a ser más calificativo, midiendo la proporción de adjetivos (token.pos_ == "ADJ") sobre el total de palabras alfabéticas.
A diferencia de las métricas anteriores, aquí se observa una diferencia leve pero constante: el promedio de adjetivos en comentarios de odio es de 0.0959, frente a un 0.0805 en comentarios sin odio. Esta tendencia se confirma en la mediana essuperior en el grupo de odio (0.058 vs 0.045).
Estos resultados sugieren que el lenguaje de odio en este corpus tiene una carga calificativa mayor. Aunque la diferencia no es lo suficientemente amplia como para ser un único factor determinante, indica que los atacantes recurren con mayor frecuencia a adjetivos para calificar peyorativamente a sus objetivos, marcando una distinción estilística frente al discurso neutro.

 

<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Característica adicional 4.</span>


<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Proporción de tokens en segunda persona gramatical</span>

In [19]:
# Incluye aquí el código generado para poder responder a tu pregunta
counts_p2 = []
for comentario in doc:
    palabras_total = [t for t in comentario if t.is_alpha]
    n_palabras = len(palabras_total)
    
    if n_palabras > 0:
        # Contamos tokens que tengan el rasgo Person=2
        n_p2 = sum(1 for t in palabras_total if '2' in t.morph.get("Person"))
        counts_p2.append(n_p2 / n_palabras)
    else:
        counts_p2.append(0.0)

# Separar por grupo
p2_odio = [p for p, v in zip(counts_p2, value) if v > 0]
p2_no_odio = [p for p, v in zip(counts_p2, value) if v == 0]

# Estadísticas
import numpy as np
print(f"ODIO - Promedio: {np.mean(p2_odio):.4f}, Mediana: {np.median(p2_odio)}, SD: {np.std(p2_odio):.4f}")
print(f"NO ODIO - Promedio: {np.mean(p2_no_odio):.4f}, Mediana: {np.median(p2_no_odio)}, SD: {np.std(p2_no_odio):.4f}")

ODIO - Promedio: 0.0174, Mediana: 0.0, SD: 0.0627
NO ODIO - Promedio: 0.0087, Mediana: 0.0, SD: 0.0397


<b>Incluye aquí, debajo de la línea, la explicación de la característica propuesta y su motivación. Incluye también una explicación del código fuente aportado.</b>
<hr><b>Característica adicional 4: Proporción de tokens en segunda persona gramatical</b> En este análisis se evaluó la hipótesis de que el discurso de odio es eminentemente directivo y se dirige a un otro a diferencia del discurso neutro. Para ello, se aprovechó el análisis morfológico que spaCy realiza durante el procesamiento del pipeline, accesible mediante el atributo token.morph. Este atributo almacena rasgos gramaticales como persona, número, tiempo verbal, género y modo. Se utilizó el método token.morph.get("Person"), el cual devuelve una lista con los valores del rasgo Person para cada token.
Se calculó la proporción de tokens con el rasgo Person=2 sobre el total de palabras alfabéticas de cada comentario, incluyendo tanto verbos conjugados en segunda persona como pronombres personales, ya que ambos son indicadores de que el mensaje va dirigido directamente a un interlocutor. Se manejó el caso de comentarios sin palabras alfabéticas asignando un valor de 0.0 para evitar divisiones por cero. Posteriormente, se separaron los resultados por grupo utilizando zip() con la lista de intensidades y se calcularon las estadísticas descriptivas con numpy.
Los resultados confirman la hipótesis de fque el promedio de uso de la segunda persona en comentarios de odio (0.0174) es exactamente el doble que en los comentarios sin odio (0.0087). Este hallazgo es el más significativo de todo el análisis estructural. Demuestra que el ataque personal directo  es una marca de la hostilidad en este corpus, doblando su frecuencia cuando existe una intención de odio. De todas las características analizadas, esta es la que presenta el mayor potencial discriminatorio para una eventual tarea de clasificación.
 

<span style="font-size: 16pt; font-weight: bold; color: #0098cd;">Reflexión final</span>

<span>Una de las utilidades de la caracterización de texto es que nos sirve como etapa de <i>feature-extraction</i> (extración de características) de cara a un posterior sistema de clasificación. Es pertinente, por tanto, reflexionar sobre la capacidad discriminatoria de cada una de las características extraídas. </span>

<span> Responde, para ello, a la siguiente pregunta.</span>

<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Reflexión final.</span>
<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">¿Es posible utilizar alguna de las características extraídas en las preguntas anteriores para determinar si un mensaje contiene odio? Justifica tu respuesta con el análisis estadístico que consideres necesario.</span>

In [20]:
# Incluye aquí el código generado para poder responder a tu pregunta
from scipy.stats import mannwhitneyu
from scipy.stats import chi2_contingency
def mannwhitney_test(df, feature, target="label", alpha=0.05):
    grupo_odio = df[df[target] == 1][feature].dropna()
    grupo_no_odio = df[df[target] == 0][feature].dropna()
    
    stat, p_value = mannwhitneyu(
        grupo_odio,
        grupo_no_odio,
        alternative="two-sided"
    )
    
    n1 = len(grupo_odio)
    n2 = len(grupo_no_odio)
    
    mean_U = n1 * n2 / 2
    std_U = np.sqrt(n1 * n2 * (n1 + n2 + 1) / 12)
    z = (stat - mean_U) / std_U
    r = z / np.sqrt(n1 + n2)
    
    if abs(r) < 0.1:
        efecto = "pequeño (casi nulo)"
    elif abs(r) < 0.3:
        efecto = "pequeño"
    elif abs(r) < 0.5:
        efecto = "medio"
    else:
        efecto = "grande"
    
    print(f"\n=== Test Mann-Whitney para {feature} ===")
    print(f"U statistic: {stat:.4f}")
    print(f"p-value: {p_value:.6f}")
    print(f"Tamaño del efecto (r): {r:.4f} -> {efecto}")
    
    if p_value < alpha:
        print("Conclusión: Diferencia estadísticamente significativa (rechazo H0)")
    else:
        print("Conclusión: No significativa (no se rechaza H0)")

label = [1 if intensidad > 0 else 0 for intensidad in value]

df = pd.DataFrame({
    "prop_p2": counts_p2,
    "prop_adj": counts_adj,
    "tiene_ner": tiene_entidades,
    "label": label
})

df.index.name = "comentario"


In [21]:
features = ["prop_p2", "prop_adj", "tiene_ner"]
for feature in features:
    mannwhitney_test(df, feature)


=== Test Mann-Whitney para prop_p2 ===
U statistic: 3663405.5000
p-value: 0.000000
Tamaño del efecto (r): 0.0268 -> pequeño (casi nulo)
Conclusión: Diferencia estadísticamente significativa (rechazo H0)

=== Test Mann-Whitney para prop_adj ===
U statistic: 3690575.5000
p-value: 0.001579
Tamaño del efecto (r): 0.0303 -> pequeño (casi nulo)
Conclusión: Diferencia estadísticamente significativa (rechazo H0)

=== Test Mann-Whitney para tiene_ner ===
U statistic: 3586916.0000
p-value: 0.042336
Tamaño del efecto (r): 0.0167 -> pequeño (casi nulo)
Conclusión: Diferencia estadísticamente significativa (rechazo H0)


In [22]:
tabla = pd.crosstab(df["label"], df["tiene_ner"])

chi2, p_value, dof, expected = chi2_contingency(tabla)

print("Tabla de contingencia:\n", tabla)
print("\nChi2:", chi2)
print("p-value:", p_value)
print("Grados de libertad:", dof)
print("\nFrecuencias esperadas:\n", expected)

Tabla de contingencia:
 tiene_ner  False  True 
label                  
0           6115   3137
1            467    281

Chi2: 3.9610671498720667
p-value: 0.04656417498556266
Grados de libertad: 1

Frecuencias esperadas:
 [[6089.6664 3162.3336]
 [ 492.3336  255.6664]]


<b>Incluye aquí, debajo de la línea, la explicación de tu respuesta</b>
<hr>
A partir de los resultados obtenidos, se concluye que algunas características presentan diferencias estadísticamente significativas entre los grupos de odio y no odio, pero su utilidad práctica como clasificadores individuales es limitada.
Las variables estructurales no mostraron capacidad discriminatoria, con diferencias mínimas entre grupos que indican que la forma superficial del texto no captura patrones asociados al discurso de odio.
Las variables lingüísticas más específicas sí evidenciaron diferencias. Para validarlas, se construyó un DataFrame con las tres características más prometedoras y una etiqueta binaria derivada de INTENSIDAD (label = [1 if intensidad > 0 else 0 for intensidad in value]). La proporción de segunda persona (prop_p2), extraída con token.morph.get("Person"), mostró el doble de frecuencia en comentarios de odio (p < 0.001). La proporción de adjetivos (prop_adj), filtrada con token.pos_ == "ADJ", también resultó significativa (p = 0.0016). Ambas se evaluaron con Mann-Whitney U (mannwhitneyu de scipy.stats), se utilizó la prueba de Mann-Whitney U debido a que las variables analizadas corresponden a proporciones y no cumplen supuestos de normalidad. La presencia de entidades NER, determinada con len(comentario.ents) > 0, mostró asociación más débil (p = 0.0466), confirmada con Chi-cuadrado de independencia (chi2_contingency, χ² = 3.96, gl = 1) al tratarse de una variable categórica binaria cuya tabla de contingencia se construyó con pd.crosstab().
Sin embargo el tamaño del efecto calculado como r = z / sqrt(n1 + n2), fue muy bajo en los tres casos (r ≈ 0.02–0.03), correspondiendo a efectos casi nulos que explican menos del 0.1% de la variabilidad entre grupos. Esto ilustra que con muestras grandes (n = 10,000), el p-value puede resultar significativo incluso para diferencias prácticamente insignificantes, por lo que debe complementarse con medidas de tamaño del efecto. Cabe señalar además el desbalance de clases del corpus solo el 7.48%  corresponde a odio frente al 92.52% neutro.
Las características capturan señales reales del discurso de odio, especialmente la segunda persona gramatical y los adjetivos, pero no son suficientemente fuertes para funcionar como clasificadores independientes. Su valor radica en su posible integración dentro de un modelo multivariado donde múltiples señales débiles se combinen para mejorar la capacidad predictiva global.
 